### PR TITLE
Refactoring AbstractRib and adding tests

### LINKED ISSUE #591: Optimize and refactor AbstractRib.java
- address comments in #590

### PR DESCRIPTION
**Pull request by progwriter**
Fixes #591

### CODE REVIEW

File: projects/batfish/src/main/java/org/batfish/bdp/AbstractRib.java [hunk 1]
```diff
@@ -85,65 +103,40 @@ public void collectRoutes(Set<R> routes) {
...
+    boolean containsRoute(R route, BitSet bits, int prefixLength, int firstUnmatchedBitIndex) {
+      // If prefix lengths match, this is the node where such route would be stored.
       if (prefixLength == _prefix.getPrefixLength()) {
         return _routes.contains(route);
+      }
+      boolean currentBit = bits.get(firstUnmatchedBitIndex);
+
+      /*
+      If prefixes don't match exactly, look at the current bit. That determines whether we look
+      left of right. As long as the child is not null, recurse.
```
Comment (dhalperi) [inline projects/batfish/src/main/java/org/batfish/bdp/AbstractRib.java]: left `or` right

Comment (dhalperi) [reply projects/batfish/src/main/java/org/batfish/bdp/AbstractRib.java]: Either use Java multiline comment form

```java
/*
 * ...
 */
```

or Java single-line block comment form:

```java
// Line 1
// Line 2
// Line 3
```

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  @Test
+  public void testSingleRouteAdd() {
+    // Test: Adding one route
+    _rib.mergeRoute(_mostGeneralRoute);
+
+    // Assertions
+    // Check routes size
+    assertThat(_rib.getRoutes().size(), is(1));
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: can use [`Matchers.hasSize`](http://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#hasSize(int)) for this, I think. Assuming it's a `Collection`, not a `Map`.

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  @Before
+  public void setupEmptyRib() {
+    _rib = new StaticRib(null);
+    _mostGeneralRoute = new StaticRoute(new Prefix("0.0.0.0/0"), Ip.ZERO, null, 0, 0);
+  }
+
+  @Test
+  public void testRibConstructor() {
+    // Test: the setupEmptyRib() fixture
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: Drop these comments -- structuring the test in a logical order with, optionally, whitespace between sections is good enough :)

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    assertThat(_rib.containsRoute(r1), is(true));
+    assertThat(_rib.containsRoute(r2), is(true));
+    // Also check route collection via getRoutes()
+    assertThat(collectedRoutes, hasItem(r1));
+    assertThat(collectedRoutes, hasItem(r2));
+  }
+
+  @Test
+  public void testMultiOverlappingRouteAdd() {
+    // Setup/Test: Add multiple routes with overlapping prefixes
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: This type of comment where useful for understanding test goals should go in the function javadoc.

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    // Assertions
+    // Check that longestPrefixMatch executes correctly
+    Set<StaticRoute> match = _rib.longestPrefixMatch(new Ip("10.1.1.1"));
+    assertThat(match.size(), equalTo(1));
+    assertThat(match, contains(routes.get(3)));
+
+    match = _rib.longestPrefixMatch(new Ip("10.1.1.2"));
+    assertThat(match.size(), equalTo(1));
+    assertThat(match, contains(routes.get(1)));
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: also test one that doesn't match (even though the rib is not empty ;).

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import java.util.List;
+import java.util.Set;
+import org.batfish.datamodel.Ip;
+import org.batfish.datamodel.Prefix;
+import org.batfish.datamodel.StaticRoute;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test the AbstractRib tree logic. To avoid worrying about route preference comparisons, testing
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: Simpler:

```java
/** Tests of {@link AbstractRib}. */
```

Then move the bit about StaticRoute to the `_rib` variable.

My general rule here is that public Javadoc is for people looking at "What does X do?" whereas private documentation is for people reading the code ("how does X work?").

[And, I don't think I care how we test AbstractRib until I'm reading the code.]

File: projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+ * with StaticRoutes and StaticRib. All static routes will be stored in the RIB, without eviction
+ * based on preference.
+ */
+public class AbstractRibTest {
+  private AbstractRib<StaticRoute> _rib;
+  private StaticRoute _mostGeneralRoute;
```
Comment (dhalperi) [inline projects/batfish/src/test/java/org/batfish/bdp/AbstractRibTest.java]: `private static final StaticRoute MOST_GENERAL_ROUTE = ...`

Since it's a constant that will never change per test.
